File: src/main/java/hiiragi283/material/network/HiiragiMessage.java
package hiiragi283.material.network;

import io.netty.buffer.ByteBuf;
import net.minecraft.nbt.NBTTagCompound;
import net.minecraft.util.math.BlockPos;
import net.minecraftforge.fml.common.network.ByteBufUtils;
import net.minecraftforge.fml.common.network.simpleimpl.IMessage;

import javax.annotation.ParametersAreNonnullByDefault;

@ParametersAreNonnullByDefault
public abstract class HiiragiMessage implements IMessage {

    public BlockPos pos = BlockPos.ORIGIN;
    public NBTTagCompound tag = new NBTTagCompound();

    public HiiragiMessage() {
    }

    public HiiragiMessage(BlockPos pos, NBTTagCompound tag) {
        this.pos = pos;
        this.tag = tag;
    }

    @Override
    public void fromBytes(ByteBuf buf) {
        int x = buf.readInt();
        int y = buf.readInt();
        int z = buf.readInt();
        pos = new BlockPos(x, y, z);
        tag = ByteBufUtils.readTag(buf);
    }

    @Override
    public void toBytes(ByteBuf buf) {
        buf.writeByte(pos.getX());
        buf.writeByte(pos.getY());
        buf.writeByte(pos.getZ());
        ByteBufUtils.writeTag(buf, tag);
    }

    public static class Client extends HiiragiMessage {

        public Client() {
            super();
        }

        public Client(BlockPos pos, NBTTagCompound tag) {
            super(pos, tag);
        }

    }

    public static class Server extends HiiragiMessage {

        public Server() {
            super();
        }

        public Server(BlockPos pos, NBTTagCompound tag) {
            super(pos, tag);
        }

    }

}

File: src/main/kotlin/hiiragi283/material/network/HiiragiMessage.kt
package hiiragi283.material.network

import hiiragi283.material.util.readBlockPos
import hiiragi283.material.util.readNBTTag
import hiiragi283.material.util.writeBlockPos
import hiiragi283.material.util.writeNBTTag
import io.netty.buffer.ByteBuf
import net.minecraft.nbt.NBTTagCompound
import net.minecraft.util.math.BlockPos

sealed class HiiragiMessage(
    override var pos: BlockPos = BlockPos.ORIGIN,
    override var tag: NBTTagCompound = NBTTagCompound()
) : IHiiragiMessage {

    override fun fromBytes(buf: ByteBuf) {
        pos = buf.readBlockPos()
        tag = buf.readNBTTag() ?: NBTTagCompound()
    }

    override fun toBytes(buf: ByteBuf) {
        buf.writeBlockPos(pos)
        buf.writeNBTTag(tag)
    }

    class Client(pos: BlockPos = BlockPos.ORIGIN, tag: NBTTagCompound = NBTTagCompound()) : HiiragiMessage(pos, tag)

    class Server(pos: BlockPos = BlockPos.ORIGIN, tag: NBTTagCompound = NBTTagCompound()) : HiiragiMessage(pos, tag)

    class Primitive(pos: BlockPos = BlockPos.ORIGIN) : HiiragiMessage(pos, NBTTagCompound())

    class SyncCount(pos: BlockPos = BlockPos.ORIGIN, var count: Int = 0): HiiragiMessage(pos, NBTTagCompound()) {

        override fun fromBytes(buf: ByteBuf) {
            super.fromBytes(buf)
            count = buf.readInt()
        }

        override fun toBytes(buf: ByteBuf) {
            super.toBytes(buf)
            buf.writeInt(count)
        }

    }

    class Player : HiiragiMessage(BlockPos.ORIGIN, NBTTagCompound())

    class Entity(
        pos: BlockPos = BlockPos.ORIGIN,
        tag: NBTTagCompound = NBTTagCompound(),
        var entityId: Int = 0
    ) : HiiragiMessage(pos, tag) {

        constructor(pos: BlockPos = BlockPos.ORIGIN, tag: NBTTagCompound = NBTTagCompound(), entity: Entity) : this(
            pos,
            tag,
            entity.entityId
        )

        override fun fromBytes(buf: ByteBuf) {
            super.fromBytes(buf)
            buf.writeInt(entityId)
        }

        override fun toBytes(buf: ByteBuf) {
            super.toBytes(buf)
            entityId = buf.readInt()
        }

    }

}